lib/common/common/src/top_k.rs

```diff
@@ -0,0 +1,151 @@
+
+    #[test]
+    fn empty_with_double_capacity() {
+        let top_k = TopK::new(3);
+        assert_eq!(top_k.len(), 0);
+        assert_eq!(top_k.elements.capacity(), 2 * 3);
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+    }
+
+    #[test]
+    fn test_top_k_under() {
+        let mut top_k = TopK::new(3);
+        top_k.push(ScoredPointOffset { score: 1.0, idx: 1 });
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+        assert_eq!(top_k.len(), 1);
+
+        top_k.push(ScoredPointOffset { score: 2.0, idx: 2 });
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+        assert_eq!(top_k.len(), 2);
+
+        let res = top_k.into_vec();
+        assert_eq!(res.len(), 2);
+        assert_eq!(res[0].score, 2.0);
+        assert_eq!(res[1].score, 1.0);
+    }
+
+    #[test]
+    fn test_top_k_over() {
+        let mut top_k = TopK::new(3);
+        top_k.push(ScoredPointOffset { score: 1.0, idx: 1 });
+        assert_eq!(top_k.len(), 1);
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+
+        top_k.push(ScoredPointOffset { score: 3.0, idx: 3 });
+        assert_eq!(top_k.len(), 2);
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+
+        top_k.push(ScoredPointOffset { score: 2.0, idx: 2 });
+        assert_eq!(top_k.len(), 3);
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+
+        top_k.push(ScoredPointOffset { score: 4.0, idx: 4 });
+        assert_eq!(top_k.len(), 4);
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+
+        let res = top_k.into_vec();
+        assert_eq!(res.len(), 3);
+        assert_eq!(res[0].score, 4.0);
+        assert_eq!(res[1].score, 3.0);
+        assert_eq!(res[2].score, 2.0);
+    }
+
+    #[test]
+    fn test_top_k_pruned() {
+        let mut top_k = TopK::new(3);
+        top_k.push(ScoredPointOffset { score: 1.0, idx: 1 });
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+        assert_eq!(top_k.len(), 1);
+
+        top_k.push(ScoredPointOffset { score: 4.0, idx: 4 });
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+        assert_eq!(top_k.len(), 2);
+
+        top_k.push(ScoredPointOffset { score: 2.0, idx: 2 });
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+        assert_eq!(top_k.len(), 3);
+
+        top_k.push(ScoredPointOffset { score: 5.0, idx: 5 });
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+        assert_eq!(top_k.len(), 4);
+
+        top_k.push(ScoredPointOffset { score: 3.0, idx: 3 });
+        assert_eq!(top_k.threshold(), ScoreType::MIN);
+        assert_eq!(top_k.len(), 5);
+
+        top_k.push(ScoredPointOffset { score: 6.0, idx: 6 });
+        assert_eq!(top_k.threshold(), 4.0);
+        assert_eq!(top_k.len(), 3);
+        assert_eq!(top_k.elements.capacity(), 6);
+
+        let res = top_k.into_vec();
+        assert_eq!(res.len(), 3);
+        assert_eq!(res[0].score, 6.0);
+        assert_eq!(res[1].score, 5.0);
+        assert_eq!(res[2].score, 4.0);
+    }
```

Comment:
Should we also include a test that has some identical scores?

Reply:
good idea https://github.com/qdrant/qdrant/pull/4037/commits/4f3307c673ddd74f083fc1157ae246d41a074a3b